### PR TITLE
CP-30614: Export Memory stats to filesystem

### PR DESCRIPTION
**Pull request by psafont**
This will allow xapi to read memory stats from the filesystem instead of depending directly on Xenctrl libs

### CODE REVIEW

**Comment by lindig** [comment]
This looks solid to me. But what does RRA stand for?

**Comment by psafont** [comment]
It stands for [Round-Robin Archive](https://github.com/xapi-project/xcp-rrd/blob/2830e698e882022fb0f5009866d28f4b3d54cdff/lib/rrd.ml#L375). Doesn't look like the correct name: I meant a file with a RRD, in a similar way to how is data from plugins being currently loaded from the filesystem.

**Comment by psafont** [comment]
I feel like this introduces quite a big behavioral change. I think some kind of unit testing is in order, any ideas on how to follow this?

I'm especially worried on life-cycle management of the writers and the termination of the daemon.
Then there is also the fragility of using strings as selectors and filenames.

I'm very open on ideas about enhancing the patchset

File: rrdd/xcp_rrdd.ml [hunk 1]
```diff
@@ -111,7 +111,7 @@ open Rrd
 let uuid_of_domid domains domid =
   try
     Rrdd_server.string_of_domain_handle
-      (List.find (fun di -> di.Xenctrl.domid = domid) domains)
+      (Lext.List.find (fun di -> di.Xenctrl.domid = domid) domains)
```
Comment (robhoes) [inline rrdd/xcp_rrdd.ml]: Most of the functions that are used from `Listext.List` below are also in the standard `List`. I only spotted one instance of `filter_map` that is special. So I think that we can drop `Lext.` from almost all of the lines below.

**Comment by robhoes** [comment]
This looks really good indeed, and it is a great example of how to split up a larger change into several logical commits (makes reviewing much easier!).

Regarding the selectors as strings, in general it may be better to use polymorphic variants. However, in this case, we still need the strings, so switching to variants only complicates things for little benefit. So I would leave it like this.

What about the case of xcp-rrdd reading back the written files and creating duplicate RRDs?

**Comment by psafont** [comment]
I'm _really_ not liking the names of the files created. I'll change the strings as well as function names, with possibly a prefix.

**Comment by psafont** [comment]
I've changed the names to be more informative to users and added a prefix to show that the files in the metrics are related to the daemon.
All ring 3 BVT and BST tests passed cleanly

**Comment by psafont** [comment]
Depends on #90

**Comment by psafont** [comment]
Now the files that are written by xcp-rrdd are propagated to the main discovery loop so they can be filtered out when the validity of files/plugins is tested:

Before:
```
Mar  5 15:27:42 nisu xcp-rrdd: [ info|nisu|4 ||rrdd_main] RRD plugin xcp-rrdd-mem_vms discovered - registering it
Mar  5 15:27:42 nisu xcp-rrdd: [ info|nisu|4 ||rrdd_main] RRD plugin xcp-rrdd-mem_host discovered - registering it
Mar  5 15:27:42 nisu xcp-rrdd: [ info|nisu|4 ||rrdd_main] RRD plugin xcp-rrdd-xenpm discovered - registering it
Mar  5 15:27:42 nisu xcp-rrdd: [ info|nisu|4 ||rrdd_main] RRD plugin xcp-rrdd-squeezed discovered - registering it
Mar  5 15:27:42 nisu xcp-rrdd: [ info|nisu|4 ||rrdd_main] RRD plugin xcp-rrdd-iostat discovered - registering it
Mar  5 15:27:47 nisu xcp-rrdd: [ info|nisu|4 ||rrdd_main] RRD plugin xapi-stats discovered - registering it
```
After:
```
Mar  5 15:55:59 nisu xcp-rrdd: [ info|nisu|4 ||rrdd_main] RRD plugin xcp-rrdd-xenpm discovered - registering it
Mar  5 15:55:59 nisu xcp-rrdd: [ info|nisu|4 ||rrdd_main] RRD plugin xcp-rrdd-squeezed discovered - registering it
Mar  5 15:55:59 nisu xcp-rrdd: [ info|nisu|4 ||rrdd_main] RRD plugin xcp-rrdd-iostat discovered - registering it
Mar  5 15:56:04 nisu xcp-rrdd: [ info|nisu|4 ||rrdd_main] RRD plugin xapi-stats discovered - registering it
```

**Comment by lindig** [comment]
Presumably there are conflicts here. Can you check this?

**Comment by psafont** [comment]
Didn't see any conflicts when rebasing.

Could be a false positive due to the merging method and that this PR depended on the PR that got merged into master.

(A rebase and merge would not have caused this)

**Comment by robhoes** [comment]
It's going in!

**Pull request merged**